src/cider/tests/CiderNewHashTableTest.cpp

```diff
@@ -0,0 +1,216 @@
+    /// Compiler cannot unroll this loop, do it manually.
+    /// (at least for floats, most likely due to the lack of -fassociative-math)
+
+    int16_t partial_sums[unroll_count]{};
+
+    const auto* end = ptr + size;
+    const auto* unrolled_end = ptr + (size / unroll_count * unroll_count);
+
+    while (ptr < unrolled_end) {
+      for (size_t i = 0; i < unroll_count; ++i)
+        partial_sums[i] += ptr[i];
+      ptr += unroll_count;
+    }
+
+    for (size_t i = 0; i < unroll_count; ++i)
+      sum += partial_sums[i];
+
+    while (ptr < end) {
+      sum += *ptr;
+      ++ptr;
+    }
+
+    count += size;
+  }
+
+  void merge(const Block& other) {
+    sum += other.sum;
+    count += other.count;
+  }
+
+  int64_t getSum() const { return sum; }
+
+  size_t getCount() const { return count; }
+
+  double getAvg() const { return (double)sum / count; }
+
+  bool operator!() const { return !count; }
+};
+
+TEST_F(CiderNewHashTableTest, aggInt8Test) {
+  using AggregatedDataWithUInt8Key = FixedImplicitZeroHashMap<int8_t, Block>;
+  AggregatedDataWithUInt8Key map;
+
+  std::vector<int8_t> keys{1, 2, 3, 4, 5};
+  std::vector<int64_t> values{10, 20, 30, 40, 50};
+
+  for (int i = 0; i < keys.size(); i++) {
+    Block& block = map[keys[i]];
+    block.add(values[i]);
+  }
+
+  for (int i = 0; i < keys.size(); i++) {
+    CHECK_EQ(map[keys[i]].getSum(), values[i]);
+    CHECK_EQ(map[keys[i]].getCount(), 1);
+  }
+
+  std::vector<int8_t> keys2{1, 2, 3, 4, 5};
+  std::vector<int64_t> values2{50, 40, 30, 20, 10};
+  for (int i = 0; i < keys2.size(); i++) {
+    Block& block = map[keys2[i]];
+    block.add(values2[i]);
+  }
+
+  for (int i = 0; i < keys.size(); i++) {
+    CHECK_EQ(map[keys[i]].getSum(), 60);
+    CHECK_EQ(map[keys[i]].getCount(), 2);
+    CHECK_EQ(map[keys[i]].getAvg(), 30);
+  }
+}
+
+TEST_F(CiderNewHashTableTest, aggInt16Test) {
+  using AggregatedDataWithUInt16Key =
+      FixedImplicitZeroHashMapWithCalculatedSize<int16_t, Block>;
+  AggregatedDataWithUInt16Key map;
+
+  std::vector<int16_t> keys{1, 2, 3, 4, 5};
+  std::vector<int64_t> values{10, 20, 30, 40, 50};
+
+  for (int i = 0; i < keys.size(); i++) {
+    Block& block = map[keys[i]];
+    block.add(values[i]);
+  }
+
+  for (int i = 0; i < keys.size(); i++) {
+    CHECK_EQ(map[keys[i]].getSum(), values[i]);
+    CHECK_EQ(map[keys[i]].getCount(), 1);
+  }
+
+  std::vector<int8_t> keys2{1, 2, 3, 4, 5};
+  std::vector<int64_t> values2{50, 40, 30, 20, 10};
+  for (int i = 0; i < keys2.size(); i++) {
+    Block& block = map[keys2[i]];
+    block.add(values2[i]);
+  }
+  for (int i = 0; i < keys.size(); i++) {
+    CHECK_EQ(map[keys[i]].getSum(), 60);
+    CHECK_EQ(map[keys[i]].getCount(), 2);
+    CHECK_EQ(map[keys[i]].getAvg(), 30);
+  }
+}
+
+TEST_F(CiderNewHashTableTest, hashTableTest) {
+  using Cell = FixedHashMapCell<int16_t, Block>;
+  FixedHashTable<int16_t, Cell, FixedHashTableStoredSize<Cell>, HashTableAllocator> fht;
```

Comment:
the second parameter have to be `FixedHashMapCell`, possible to be some simple type for this ut/example?

Reply:
@Deegue please address this comment in later PR.

Reply:
Users will not need to care about Cell like `FixedHashMapCell<key_type, value_type>` or `HashMapCell<key_type, value_type>` since it's related to the HashTable implementation decided by the chooser. Users only need to pass down `key_type` and `value_type`.

Reply:
what should I do if I want a simple hash map which `key_type` is Int32 and `value_type` is Int64

Reply:
When using `Aggregator`, you can set via constructor. But for hash map, it cannot avoid setting a `Cell`. It's not universal for simple hash map.

Exposing `Aggregator`, `Join` may be better than simple `hashtable` to users for easier integration.

Reply:
if so, `Cell` will be part of `hashtable` API, please call out that.